contributor-docs/adrs/adr-019-responsive-values.md

```diff
@@ -0,0 +1,159 @@
+authors should use the following patterns:
+
+```tsx
+type GapScale = 'none' | 'condensed' | 'normal' | 'spacious'
+
+type StackProps = React.PropsWithChildren<{
+  // Responsive props should be a union of the value of the prop plus
+  // `ResponsiveValue` over that value
+  gap?: GapScale | ResponsiveValue<GapScale>
+}>
+
+function Stack({children, gap}: StackProps) {
+  // The `getResponsiveAttributes` helper is useful for adding the corresponding
+  // `data-*` attributes to the container for the compnoent.
+  //
+  // In this case, it will add `data-gap` and `data-gap-{breakpoint}` attributes
+  // with the corresponding value set
+  return <StyledStack {...getResponsiveAttributes('gap', gap)}>{children}</StyledStack>
+}
+
+const StyledStack = styled.div`
+  &[data-gap='none'],
+  &[data-gap-narrow='none'] {
+    /* ... */
+  }
+
+  &[data-gap='condensed'],
+  &[data-gap-narrow='condensed'] {
+    /* ... */
+  }
+
+  /* ... */
+
+  // @custom-media --viewportRange-regular
+  @media (min-width: 768px) {
+    &[data-gap-regular='none'] {
+      /* ... */
+    }
+
+    &[data-gap-regular='condensed'] {
+      /* ... */
+    }
+
+    /* ... */
+  }
+
+  // @custom-media --viewportRange-wide
+  @media (min-width: 1400px) {
+    &[data-gap-wide='none'] {
+      /* ... */
+    }
+
+    &[data-gap-wide='condensed'] {
+      /* ... */
+    }
+
+    /* ... */
+  }
+`
+```
+
+By default, this approach uses the following conventions:
+
+- Props which may have responsive values should have a value that is
+  serializable, ideally as a `string` or `boolean`
+- Props will be provided on the component as a data attribute, using the format:
+  `data-{propName}`
+- For responsive values, the breakpoint name will be appended to the data
+  attribute name specified above. For example: `data-{propName}-{breakpoint}`
+
+Data attributes will act as a bridge between JavaScript and CSS. In CSS, authors
+should use these data attributes to apply styles based on the value of the
+attribute per-breakpoint.
+
+> [!NOTE]
+> There is a utility named `getResponsiveAttributes` that can be used to generate
+> the corresponding `data-*` attributes for the props which may have responsive
+> values.
+
+### Storybook
+
+When authoring playgrounds for components that have responsive values, use
+the following strategy for writing controls:
+
+- Provide a control for the plain value of the prop, for example `gap`
+- For each supported responsive value, provide a control with the name of the
+  breakpoint appended to the prop name, for example:
+  - `gapNarrow`
+  - `gapNormal`
+  - `gapWide`
+- Use the `getRespnosiveControlValues` helper to resolve between the plain and
+  responsive value for the prop. By default, this helper will pass along a
+  responsive value if one is defined. If none are defined, the plain value will
+  be used
+
+### Impact
+
+- Use of the `useResponsiveValue` hook should be deprecated and discouraged
```

Comment:
> Authors should use data attributes to bridge between responsive values in
JavaScript and the corresponding styles in CSS.

I agree with the overall direction and decision. That's why approving the PR.

> - Use of the `useResponsiveValue` hook should be deprecated and discouraged

I don't want to commit to the deprecation just yet because I have an example where a CSS only solution would be more complex to maintain: 

In SelectPanel2, which composes other components, we [set the button size based on the responsive value](https://github.com/primer/react/blob/main/packages/react/src/drafts/SelectPanel2/SelectPanel.tsx#L480)

contrived source code: 

```tsx
const responsiveButtonSizes: ResponsiveValue<'small' | 'medium'> = {narrow: 'medium', regular: 'small'}

const SelectPanelFooter = ({...props}) => {
  const buttonSize = useResponsiveValue(responsiveButtonSizes, 'small')

  return (
      <div>
        <div>{props.children}</div>
        <Button type="button" size={buttonSize}}>Cancel</Button>
        <Button type="submit" size={buttonSize} variant="primary">Save</Button> 
      </div>
  )
}
```